### PR TITLE
improve moerr to support mo-error's unified format and reporting

### PR DESCRIPTION
**Pull request by xzxiong**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #3435, 

## What this PR does / why we need it:
mo has its own error code/message, and need to adapt mysql error code.
- define `moErrorMsgItem{errorCode, mysqlErrorCode, errorMsgOrFormat}` to handle `moErrorCode->mysqlErrorCode` mapping rules
- define `New(errorCoe, args)` and `NewWithContext(context.Context, errorCode, args)` to handle error reporting
- modify reference of `moerr.ErrTxnClosed` error, as an example of `moerr.New(...)`
- predefine `Group 6: sql error` as sql error, which normally can be converted to mysql error code.

### CODE REVIEW

File: pkg/common/moerr/error.go [hunk 1]
```diff
@@ -57,24 +63,118 @@ const (
...
+		}
+	} else {
+		err = &Error{
+			MysqlErrCode: item.mysqlErrorCode,
+			Code:         code,
+			ErrorCode:    item.errorCode,
+			Message:      fmt.Sprintf(item.errorMsgOrFormat, args...),
+		}
+	}
+	_ = errors.WithContextWithDepth(ctx, err, depth+1)
```
Comment (fengttt) [inline pkg/common/moerr/error.go:169]: Minor, will we ever call newWithDepth with depth not equals to 1?

**Review by fengttt — APPROVED** [review]
LGTM, thank u.

File: pkg/common/moerr/error.go (hunk 1 above)
Comment (xzxiong) [reply pkg/common/moerr/error.go:169]: got it, obviously no.

**Review by zhangxu19830126 — APPROVED** [review]

**Review by ouyuanning — APPROVED** [review]

**Review by nnsgmsone — APPROVED** [review]

**Comment by xzxiong** [comment]
@mergify refresh

**Comment by xzxiong** [comment]
@mergify refresh

**Comment by xzxiong** [comment]
@mergify refresh

**Comment by xzxiong** [comment]
@mergify refresh

**Comment by xzxiong** [comment]
@merigfy refresh

**Comment by xzxiong** [comment]
@mergify refresh

**Comment by xzxiong** [comment]
@ mergify refresh

**Comment by xzxiong** [comment]
@mergify refresh

**Pull request merged**